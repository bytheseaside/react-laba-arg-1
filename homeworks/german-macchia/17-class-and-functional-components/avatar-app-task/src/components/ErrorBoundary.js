import React, { Component } from "react";

class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }
  
  componentDidCatch(error, errorInfo) {
    this.setState({
      hasError: true,
      error: error,
      errorInfo: errorInfo,
    });
  }

  render() {
    if (this.state.hasError) {
      return (
        <div className="errorBoundary">
          <div className="errorBoundary__info">
            <h2>Something went wrong</h2>
            <details style={{ whiteSpace: "pre-wrap" }}>
              <br />
              {this.state.error.toString()}
            </details>
          </div>
        </div>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;
