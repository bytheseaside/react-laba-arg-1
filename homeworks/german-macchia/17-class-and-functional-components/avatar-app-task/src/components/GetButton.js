import loadingCircle from "../assets/loading_circle.svg";
import plusIcon from "../assets/plus_icon.svg";
import React, { Component } from "react";

class GetButton extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: false,
      pushPhoto: props.pushPhoto,
    };
  }

  //when pushed, will trigger to father's component App pushPhoto Function
  triggerPushPhoto = async () => {
    this.setLoading(true);
    this.state.pushPhoto();
  };

  setLoading = (value) => this.setState(() => ({ loading: value }));

  render() {
    return (
      <button className="square_button" onClick={this.triggerPushPhoto}>
        {this.state.loading ? (
          <img
            className="square_button__loading"
            src={loadingCircle}
            alt="loading Circle"
          />
        ) : (
          <img className="square_button__icon" src={plusIcon} alt="plus_icon" />
        )}
      </button>
    );
  }
}

export default GetButton;
