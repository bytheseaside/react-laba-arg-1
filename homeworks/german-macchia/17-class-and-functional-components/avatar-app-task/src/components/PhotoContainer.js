import React, { PureComponent } from "react";
import fetchPhotos from "../fetchPhotos";

class PhotoContainer extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      url: props.url,
      name: props.name,
      id: props.id,
      error: false,
      errorMessage: "",
      isReloading: false,
    };
  }

  // reload new data for this single container
  reload = async () => {
    this.setState(() => ({
      isReloading: true,
    }));
    try {
      const quantity = 1;
      const response = await fetchPhotos(quantity);
      const [photo] = await response.json();
      this.setState(() => ({
        id: photo.first_name + photo.id,
        url: photo.url,
        name: `${photo.first_name} ${photo.last_name}`,
      }));
    } catch (error) {
      this.setState({
        error: true,
        errorMessage: error.message,
      });
    } finally {
      this.setState({ isReloading: false });
    }
  };

  render() {
    if (this.state.error) {
      throw new Error(this.state.errorMessage);
    }
    return (
      <div className="container" onClick={() => this.reload()}>
        <img
          className="container__profile_image"
          src={this.state.url}
          alt={this.state.name}
        />
        <div
          className={
            this.state.isReloading
              ? "container__image_box container__image_box--animate"
              : "container__image_box"
          }
        />
      </div>
    );
  }
}

export default PhotoContainer;
