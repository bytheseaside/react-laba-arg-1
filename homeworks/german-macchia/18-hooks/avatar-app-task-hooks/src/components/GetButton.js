import loadingCircle from "../assets/loading_circle.svg";
import plusIcon from "../assets/plus_icon.svg";
import React from "react";

export const GetButton = React.memo(({ handleNewPhoto, loading }) => {
  return (
    <button className="square_button" onClick={handleNewPhoto}>
      {loading ? (
        <img
          className="square_button__loading"
          src={loadingCircle}
          alt="loading Circle"
        />
      ) : (
        <img className="square_button__icon" src={plusIcon} alt="plus_icon" />
      )}
    </button>
  );
});
