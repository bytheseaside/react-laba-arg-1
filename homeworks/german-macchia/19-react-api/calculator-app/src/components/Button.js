import React from "react";

export const Button = React.memo(
  ({ boxStyle, innerHtml, value, handleOnPushKey }) => {
    return (
      <div className={boxStyle} onClick={() => handleOnPushKey(value)}>
        <div className="center">{innerHtml}</div>
      </div>
    );
  }
);
