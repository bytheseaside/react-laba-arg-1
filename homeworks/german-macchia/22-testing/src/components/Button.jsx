import React from "react";

export const Button = React.memo(
  ({ boxStyle, innerHtml, value, handleOnPushKey }) => {
    return (
      <div
        className={boxStyle}
        aria-label="command-key"
        onClick={() => handleOnPushKey(value)}
      >
        <div className="center" data-testid="innerText">
          {innerHtml}
        </div>
      </div>
    );
  }
);
