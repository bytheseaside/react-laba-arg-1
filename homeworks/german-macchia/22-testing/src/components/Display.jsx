import React, { useEffect, useState } from "react";
import isEmpty from "../helpers/isEmpty";
import backspace from "../helpers/backspace";
import resolve from "../helpers/resolve";

export const Display = ({ command, press }) => {
  const [resolution, setResolution] = useState("");
  const [calculation, setCalculation] = useState("");

  //Clear calculation field, if it's empty, sets resolution field to 0
  const clear = () => {
    if (isEmpty(calculation)) {
      setResolution(0);
    }
    setCalculation("");
  };

  const handleSolution = () => {
    const result = resolve(calculation, resolution);
    setResolution(result);
    clear();
  };

  useEffect(() => {
    //If last char in calculation string & command are a non digit value
    //and also command is NOT a 'equal, delete, or clear' command value
    //Then, replace that last operator command
    if (
      /\D/g.test(calculation.slice(-1)) &&
      /\D/g.test(command) &&
      /[^CD=]/g.test(command)
    ) {
      setCalculation(calculation.slice(0, -1) + command);
      return;
    }

    //Swich command of 'equal, delete, or clear' if exists
    switch (command) {
      case "=":
        handleSolution();
        return;
      case "D":
        setCalculation(backspace(calculation));
        return;
      case "C":
        clear();
        return;
      default:
        break;
    }

    //set max 17 char length in calculation
    if (calculation.length < 17) {
      setCalculation(calculation + command);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [command, press]);

  return (
    <div data-testid="calculation-container" className="display">
      <h1 data-testid="calculation-display">{calculation}</h1>
      <div
        data-testid="resolution-container"
        className={
          resolution.toString().length > 9
            ? "display__resolution display__resolution--small"
            : "display__resolution"
        }
      >
        <h1 data-testid="resolution-display">
          {resolution.toString().slice(0, 12)}
        </h1>
      </div>
    </div>
  );
};
