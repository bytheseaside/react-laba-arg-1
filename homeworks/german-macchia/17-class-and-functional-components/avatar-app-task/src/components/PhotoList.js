import PhotoContainer from "./PhotoContainer";
import React, { Component } from "react";
import fetchPhotos from "../fetchPhotos";
import ErrorBoundary from "./ErrorBoundary";

class PhotoList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      photos: [],
      refreshList: this.refreshList,
      error: false,
      errorMessage: "",
    };
  }

  //fetch from api, receives quantity, returns array
  getPhotos = async (quantity) => {
    const response = await fetchPhotos(quantity);
    if (response.status !== 200) {
      this.setState(() => ({
        error: true,
        errorMessage: `STATUS ${response.status}, ${response.statusText}`,
      }));
    } else {
      const json = await response.json();
      return json.map((photo) => {
        return {
          id: photo.first_name + photo.id,
          url: photo.url,
          name: `${photo.first_name} ${photo.last_name}`,
        };
      });
    }
  };

  //Set length from array, call getPhoto with same quantity
  refreshList = async () => {
    const statePhotosLength = this.state.photos.length;
    if (statePhotosLength) {
      const newPhotos = await this.getPhotos(statePhotosLength);
      this.setState({ photos: newPhotos });
    }
  };

  //Push single photo to array
  addPhoto = async () => {
    const quantity = 1;
    const photo = await this.getPhotos(quantity);
    if (photo) {
      this.setState((state) => ({ photos: [...state.photos, ...photo] }));
    }
  };

  //Returns array of one Photocontainer from object in array
  displayPhotos = () =>
    this.state.photos.map((photo) => {
      return (
        <ErrorBoundary key={photo.name + photo.id}>
          <PhotoContainer url={photo.url} name={photo.name} id={photo.id} />
        </ErrorBoundary>
      );
    });

  render() {
    if (this.state.error) {
      throw new Error(this.state.errorMessage);
    }
    return <>{this.displayPhotos()}</>;
  }
}

export default PhotoList;
