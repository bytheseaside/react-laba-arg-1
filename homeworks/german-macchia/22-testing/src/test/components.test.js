/**
 * @jest-environment jsdom
 */
import { render, screen } from "@testing-library/react";
import { Button } from "../components/Button";
import { Display } from "../components/Display";

describe("Test in <Button />", () => {
  const props = {
    boxStyle: "button border-box corner--left",
    innerHtml: "AC",
    value: 0,
  };

  const button = (
    <Button
      innerHtml={props.innerHtml}
      boxStyle={props.boxStyle}
      value={props.value}
    />
  );

  test("should be a match between render and snapshot", () => {
    render(button);
    expect(screen.container).toMatchSnapshot();
  });

  test("should match text in innerText data value", () => {
    render(button);
    expect(screen.getByTestId("innerText").innerHTML).toBe(props.innerHtml);
  });
});

describe("Test in <Display />", () => {
  const display = <Display />;

  test("should be a match between render and snapshot", () => {
    render(display);
    expect(screen.container).toMatchSnapshot();
  });

  test("should contain a calculation container and display", () => {
    render(display);
    expect(screen.getByTestId("calculation-container")).toBeTruthy();
    expect(screen.getByTestId("calculation-display")).toBeTruthy();
  });

  test("should contain a resolution container and display", () => {
    render(display);
    expect(screen.getByTestId("resolution-container")).toBeTruthy();
    expect(screen.getByTestId("resolution-display")).toBeTruthy();
  });
});
