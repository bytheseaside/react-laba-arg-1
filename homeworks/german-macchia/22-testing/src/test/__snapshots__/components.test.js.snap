// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <Button /> should be a match between render and snapshot 1`] = `undefined`;

exports[`Test in <Display /> should be a match between render and snapshot 1`] = `undefined`;
